File: code/py/ch04/04.01.py
x = [165, 170, 175, 180, 185]
import numpy as np
np.mean(x) # リストの場合
#> 175.0

x = np.array([165, 170, 175, 180, 185])
x.mean() # アレイの場合（np.mean(x)も可）
#> 175.0

import pandas as pd
x = pd.Series([165, 170, 175, 180, 185])
x.mean() # シリーズの場合（np.mean(x)も可）
#> 175.0

n = len(x) # 標本の大きさ
sum(x) / n
#> 

y = [173, 174, 175, 176, 177]
np.mean(y)
#> 175.0

np.var(x) # xの分散
#> 50.0

np.var(y) # yの分散
#> 2.0

sum((x - np.mean(x))**2) / n # 分散
#> 50.0

np.std(x) # xの標準偏差
#> 7.0710678118654755

np.std(y) # yの標準偏差
#> 1.4142135623730951

np.var(x)**0.5 # xの標準偏差
#> 7.0710678118654755

import pandas as pd
s = pd.Series(x)
s.describe()
#> count      5.000000 （データ数）
#> mean     175.000000 （平均）
#> std        7.905694 （標準偏差）
#> min      165.000000 （最小値）
#> 25%      170.000000 （第1四分位数）
#> 50%      175.000000 （中央値）
#> 75%      180.000000 （第2四分位数）
#> max      185.000000 （最大値）
#> dtype: float64

# s.describe()で計算済み

my_df = pd.DataFrame({
    'english': [60, 90, 70,  90],
    'math':    [70, 80, 90, 100]})
my_df.describe()
#>        english        math
#> count      4.0    4.000000
#> mean      77.5   85.000000
#> std       15.0   12.909944
# 以下省略

import numpy as np
x = [165, 170, 175, 180, 185]
n = len(x)

np.var(x)
# あるいは
np.var(x, ddof=0)
#> 50.0

np.var(x, ddof=1)
# あるいは
np.var(x) * n / (n - 1)
#> 62.5

# 不偏分散の平方根
np.std(x, ddof=1)
# あるいは
np.sqrt(n / (n - 1)) * np.std(x)
#> 7.905694150420949

# 標準偏差'
np.std(x)
# あるいは
np.std(x, ddof=0)
#> 7.0710678118654755

np.std(x, ddof = 1) / len(x)**0.5
#> 3.5355339059327373

import pandas as pd

my_df = pd.DataFrame({
    'name':    ['A', 'B', 'C', 'D'],
    'english': [ 60,  90,  70,  90],
    'math':    [ 70,  80,  90, 100],
    'gender':  ['f', 'm', 'm', 'f']})

my_df['english'].var()
# あるいは
import numpy as np
np.var(my_df['english'], ddof=1)

#> 225.0

my_df.var()
# あるいは
my_df.agg('var')

#> english    225.000000
#> math       166.666667
#> dtype: float64

my_df.iloc[:, [1,2]].agg(
    lambda x: np.std(x, ddof=1) /
    len(x)**0.5)
#> english    7.500000
#> math       6.454972
#> dtype: float64

my_df.describe()
#>        english        math
#> count      4.0    4.000000
#> mean      77.5   85.000000
#> std       15.0   12.909944
#> min       60.0   70.000000
#> 25%       67.5   77.500000
#> 50%       80.0   85.000000
#> 75%       90.0   92.500000
#> max       90.0  100.000000





my_df.groupby('gender').count()
# あるいは
my_df.groupby('gender').agg(len)

#>         name  english  math
#> gender
#> f          2        2     2
#> m          2        2     2

my_df.groupby('gender').mean()
# あるいは
my_df.groupby('gender').agg('mean')

#>         english  math
#> gender
#> f          75.0  85.0
#> m          80.0  85.0

my_df.groupby('gender').agg(
    lambda x: x.std() /
    len(x)**0.5)
#>         english  math
#> gender
#> f            15    15
#> m            10     5

File: code/R/ch04/04.01.R
### 4.1.1 平均・分散・標準偏差

x <- c(165, 170, 175, 180, 185)

mean(x) # 平均（その1）
#> [1] 175

sum(x) / length(x) # 平均（その2）
#> [1] 175

x - mean(x)
#> [1] -10  -5   0   5  10

mean(x - mean(x))
#> [1] 0

mean((x - mean(x))^2)
#> [1] 50

sqrt(mean((x - mean(x))^2))
#> [1] 7.071068

y <- c(173, 174, 175, 176, 177)

mean(y)
#> [1] 175 # 平均

sqrt(mean((y - mean(y))^2))
#> [1] 1.414214 # 標準偏差

var(x) # 分散
#> [1] 62.5

sd(x) # 標準偏差
#> [1] 7.905694

var(x) # 不偏分散
#> [1] 62.5

sd(x) # 平方根
#> [1] 7.905694

n <- length(x)
var(x) * (n - 1) / n # 不偏でない分散
#> [1] 50

sd(x) * sqrt((n - 1) / n) # 平方根
#> [1] 7.071068

psych::describe(x)
#>    vars n mean   sd ...
#> X1    1 5  175 7.91 ...

x <- 1:9
quantile(x)
#>   0%  25%  50%  75% 100% 
#>    1    3    5    7    9 

my_df <- data.frame(
  english = c( 60,  90,  70,  90),
  math    = c( 70,  80,  90, 100))
psych::describe(my_df)

#>         vars n mean    sd ...
#> english    1 4 77.5 15.00 ...
#> math       2 4 85.0 12.91 ...

### 4.1.2 可視化

head(iris)
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1          5.1         3.5          1.4         0.2  setosa
#> 2          4.9         3.0          1.4         0.2  setosa
#> 3          4.7         3.2          1.3         0.2  setosa
#> 4          4.6         3.1          1.5         0.2  setosa
#> 5          5.0         3.6          1.4         0.2  setosa
#> 6          5.4         3.9          1.7         0.4  setosa

#### 4.1.2.1 ヒストグラム

hist(iris$Sepal.Length) # ヒストグラム

x <- c(1, 2, 3)
hist(x, breaks = 2) # 階級数は2

#### 4.1.2.2 箱ひげ図

boxplot(iris[, -5])

psych::describe(iris)
#>              vars   n mean   sd median trimmed  mad min max range  skew kurtosis   se
#> Sepal.Length    1 150 5.84 0.83   5.80    5.81 1.04 4.3 7.9   3.6  0.31    -0.61 0.07
#> Sepal.Width     2 150 3.06 0.44   3.00    3.04 0.44 2.0 4.4   2.4  0.31     0.14 0.04
#> Petal.Length    3 150 3.76 1.77   4.35    3.76 1.85 1.0 6.9   5.9 -0.27    -1.42 0.14
#> Petal.Width     4 150 1.20 0.76   1.30    1.18 1.04 0.1 2.5   2.4 -0.10    -1.36 0.06
#> Species*        5 150 2.00 0.82   2.00    2.00 1.48 1.0 3.0   2.0  0.00    -1.52 0.07

boxplot(x = iris,
  formula = Sepal.Length ~ Species)

iris %>% 
  group_by(Species) %>%
  summarise(mean = mean(Sepal.Length),
            median = median(Sepal.Length))
#>   Species     mean median
#>   <fct>      <dbl>  <dbl>
#> 1 setosa      5.01    5  
#> 2 versicolor  5.94    5.9
#> 3 virginica   6.59    6.5

#### 4.1.2.3 散布図

plot(iris$Sepal.Length,
     iris$Sepal.Width)

#### 4.1.2.4 関数のグラフ

curve(x^3 - x, -2, 2)

# あるいは
f <- function(x) { x^3 - x }
data.frame(x = c(-2, 2)) %>%
  ggplot(aes(x)) +
  stat_function(fun = f)

### 4.1.3 ggplot2 (R)

データフレーム %>%
  ggplot(aes(x = 横軸に使うもの, y = 縦軸に使うもの)) +
  幾何オブジェクト1(オプション) +
  幾何オブジェクト2(オプション) + ...

# ヒストグラム
g1 <- iris %>%
  ggplot(aes(x = Sepal.Length)) +
  geom_histogram(bins = 8)

# 箱ひげ図（その1）
tmp <- iris %>% pivot_longer(-Species) # 整然データ
g2 <- tmp %>% 
  ggplot(aes(x = name, y = value)) +
  geom_boxplot()

# 箱ひげ図（その2）
g3 <- iris %>%
  ggplot(aes(x = Species, y = Sepal.Length)) +
  geom_boxplot()

# 散布図
g4 <- iris %>%
  ggplot(aes(x = Sepal.Length, y = Sepal.Width)) +
  geom_point()

# 並べて描画（列数は2）
gridExtra::grid.arrange(g1, g2, g3, g4, ncol = 2)

